xarray/coding/times.py

```diff
@@ -253,7 +253,11 @@ def infer_datetime_units(dates):
     else:
         reference_date = dates[0] if len(dates) > 0 else '1970-01-01'
         reference_date = format_cftime_datetime(reference_date)
-    unique_timedeltas = np.unique(np.diff(dates)).astype('timedelta64[ns]')
+    unique_timedeltas = np.unique(np.diff(dates))
+    if unique_timedeltas.dtype == np.dtype('O'):
+        # Convert to np.timedelta64 objects using pandas to work around a
+        # NumPy casting bug: https://github.com/numpy/numpy/issues/11096
+        unique_timedeltas = pd.to_timedelta(unique_timedeltas).values
```

Comment:
use `pd.to_timedelta(unique_timedeltas, box=False)` instead of `pd.to_timedelta(unique_timedeltas).values`.